packages/matter.js/src/mdns/MdnsServer.ts

```diff
@@ -77,28 +91,74 @@ export class MdnsServer {
-                    DnsCodec.encode({
-                        messageType: DnsMessageType.Response,
-                        transactionId,
-                        answers,
-                        additionalRecords,
-                    }),
-                    netInterface,
-                )
-                .catch(() => {
-                    // ignore because already caught in UdpMulticastServer
-                });
+            // Only send additional records if the query is not for A or AAAA records
+            let additionalRecords =
+                queries.find(
+                    query => query.recordType !== DnsRecordType.A && query.recordType !== DnsRecordType.AAAA,
+                ) !== undefined
+                    ? portRecords.filter(record => !answers.includes(record) && record.recordType !== DnsRecordType.PTR)
+                    : [];
+            if (knownAnswers.length > 0) {
+                for (const knownAnswersRecord of knownAnswers) {
+                    answers = answers.filter(record => !isDeepEqual(record, knownAnswersRecord));
+                    if (answers.length === 0) break; // Nothing to send
+                }
+                if (answers.length === 0) continue; // Nothing to send
+                if (additionalRecords.length > 0) {
+                    for (const knownAnswersRecord of knownAnswers) {
+                        additionalRecords = additionalRecords.filter(
+                            record => !isDeepEqual(record, knownAnswersRecord),
+                        );
+                    }
+                }
+            }
+
+            const now = Time.nowMs();
+            let uniCastResponse = queries.filter(query => !query.uniCastResponse).length === 0;
+            const answersTimeSinceLastMultiCast = answers.map(answer => ({
+                timeSinceLastMultiCast:
+                    now - (this.recordLastSentAsMulticastAnswer.get(this.buildDnsRecordKey(answer, netInterface)) ?? 0),
+                ttl: answer.ttl,
+            }));
+            if (
+                uniCastResponse &&
+                answersTimeSinceLastMultiCast.filter(
+                    ({ timeSinceLastMultiCast, ttl }) => timeSinceLastMultiCast > (ttl / 4) * 1000,
+                ).length > 0
+            ) {
+                // If the query is for unicast response, still send as multicast if they were last sent as multicast longer then 1/4 of their ttl
+                uniCastResponse = false;
+            }
+            if (!uniCastResponse) {
+                answers = answers.filter(
+                    (_, index) => answersTimeSinceLastMultiCast[index].timeSinceLastMultiCast > 1000,
+                );
+                if (answers.length === 0) continue; // Nothing to send
+
+                answers.forEach(answer =>
+                    this.recordLastSentAsMulticastAnswer.set(this.buildDnsRecordKey(answer, netInterface), now),
+                );
+            }
+
+            this.sendRecords(
```

Comment:
In general even if we want stuff to execute asynchronously we should probably be keeping track of the promises and doing an await during teardown, before repeating the operation, on cancel, etc.

Otherwise I worry hangs and/or other unexpected behavior could lead to resource leaks and generally difficult-to-diagnose issues.

Reply:
see above. on shutdown we close all underlaying network sockets and such which should end all these gracefully anyway - additional to "I know no wway to cancel a Promise", beside what said ... closing/ending underlaying socket

Reply:
I meant canceling the operation, not via the Promise interface.  Closing the underlying network sockets seems like the right thing to do.  Then the async networking code notes termination and returns, the promise is fulfilled and the termination code exits cleanly.

Ideally we'd then have a timeout promise in a Promise.any() that prints a loud "network logic failed to terminate" if the cleanup code didn't exit cleanly in the allotted time.

Point though is that we've got a handle on ongoing async operations so they can't hang without our noticing.

I could take on as a TODO but need to note the places in the code where we're letting an async operation "dangle".

Reply:
As told in discord I would more think about a general MDNS sending  timeout thing on socket level. Roadmap todo added

Reply:
Resolving per conversation in Discord